### PR TITLE
[Core] Update to outlines >= 0.1.8

### LINKED ISSUE #3794: [Feature]: Make `outlines` dependency optional
### 🚀 The feature, motivation and pitch

I'm using a newer version of `outlines` than v0.0.34, and my application needs the fixes implemented in newer versions of that package. It would be great if `vllm` could make its structured generation dependencies optional, as I'm not using the internal `vllm` structured generation, but rather using the logits processors from `outlines` directly.

### Alternatives

_No response_

### Additional context

This prevents me from updating to `vllm==0.4.0`.

### LINKED ISSUE #10489: [Feature]: Support outlines versions > v0.1.0
### 🚀 The feature, motivation and pitch

Outlines (used for guided decoding) has had a minor upgrade on 7th of Oct (ref below).
This has improved their internal workings and should have a good impact on the performance downstream (on vllm).

At the moment the version used by vllm is pinned but we have successfully spinned up vllm ourselves internally with just upgrading this dependency (there are no _relevant_ breaking changes for vllm) and it works out of the box.

https://github.com/dottxt-ai/outlines/releases/tag/0.1.0

### Alternatives

_No response_

### Additional context

_No response_

### Before submitting a new issue...

- [X] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by russellb**
This PR updates to the latest release of `outlines` that works with vllm.

It is a draft while we wait for 0.1.8 to be on pypi.

FIX https://github.com/vllm-project/vllm/issues/3794 
FIX https://github.com/vllm-project/vllm/issues/10489

279ccc9dd [Core] Update to outlines >= 0.1.8

commit 279ccc9ddd20094b8ac94e245718d534515ac6b9
Author: Russell Bryant <rbryant@redhat.com>
Date:   Thu Nov 21 21:25:22 2024 +0000

    [Core] Update to outlines >= 0.1.8
    
    0.1.x prior to 0.1.8 + outlines-core 0.1.18 had issues with
    serialization that broke vllm integration.
    
    Also change our code slightly to account for an API change in
    outlines.
    
    Signed-off-by: Russell Bryant <rbryant@redhat.com>

### CODE REVIEW

**Comment by russellb** [comment]
converted back to a draft because I don't have `0.1.5` working locally, yet ...

**Comment by joennlae** [comment]
I introduce this PR to outlines:

https://github.com/dottxt-ai/outlines-core/pull/99

The caching issue (or not being able to pickle) also affects the multiprocessing-based engine in `vllm` as we need to be able to pickle the `RegexGuide`. I introduce the pickling capability with the above PR.

**Comment by markmc** [comment]
> I introduce this PR to outlines:
> 
> [dottxt-ai/outlines-core#99](https://github.com/dottxt-ai/outlines-core/pull/99)
> 
> The caching issue (or not being able to pickle) also affects the multiprocessing-based engine in `vllm` as we need to be able to pickle the `RegexGuide`. I introduce the pickling capability with the above PR.

Thank you @joennlae - this works for me

Here are some performance measurements with older outlines, newer outlines with your patch but no caching, newer outlines with your patch and caching reinstated:

```
   commit f9310cbd0c1109c4f22cf9f1dc615b2d08f06408 from main
   with	pull/10557/head:xuechendi-benchmark-structured-output commit 41966603
   outlines==0.0.46

   $ python benchmarks/benchmark_guided.py --async-engine --model meta-llama/Llama-3.2-1B-Instruct --output-len 2048 --num-prompts 10 --save-results
   Throughput: 0.14 requests/s, 361.06 total tokens/s, 281.59 output tokens/s Correct rate is 100.0 %
   Throughput: 0.16 requests/s, 414.12 total tokens/s, 322.97 output tokens/s Correct rate is 100.0 %
   Throughput: 0.16 requests/s, 414.91 total tokens/s, 323.59 output tokens/s Correct rate is 100.0 %

   commit b8c2895a2cbe249e86713615c9ed3ab132812b08 from russellb/outlines-0.1.4
   with pull/10557/head:xuechendi-benchmark-structured-output commit 41966603
   outlines==0.1.7 (from PyPi)
   outlines_core 412ef296392a0814a5490ccc15080e79f98cd411 from 44ai-labs/serialization "release build" (pip install .)

   $ python benchmarks/benchmark_guided.py --async-engine --model meta-llama/Llama-3.2-1B-Instruct --output-len 2048 --num-prompts 10 --save-results
   Throughput: 0.06 requests/s, 167.18 total tokens/s, 130.38 output tokens/s Correct rate is 100.0 %
   Throughput: 0.06 requests/s, 167.48 total tokens/s, 130.61 output tokens/s Correct rate is 100.0 %
   Throughput: 0.07 requests/s, 171.11 total tokens/s, 133.45 output tokens/s Correct rate is 100.0 %

   same	as above, with "Stop using outlines.caching.cache" reverted
   Throughput: 0.16 requests/s, 410.16 total tokens/s, 319.88 output tokens/s Correct rate is 100.0 %
   Throughput: 0.15 requests/s, 406.16 total tokens/s, 316.76 output tokens/s Correct rate is 100.0 %
   Throughput: 0.16 requests/s, 408.68 total tokens/s, 318.73 output tokens/s Correct rate is 100.0 %
```

**Comment by russellb** [comment]
I think this will be ready once outlines 0.1.8 is available on pypi.

**Comment by rlouf** [comment]
It is !

**Review by mgoin — APPROVED** [review]
LGTM pending green CI!

**Review by aarnphm — APPROVED** [review]

**Review by aarnphm — APPROVED** [review]
